### PR TITLE
fix https request on http proxy problem

### PR DESCRIPTION
**Pull request by stoneChen**
This was a known problem as https://github.com/axios/axios/issues/925.
I found that https://github.com/request/request does not have this problem. So I debugged both of tow libs, and post this fix.
The key is that, http and https request should not be handled the same way. The https request should use tunnel instead.
See https://github.com/request/request#proxies.

### CODE REVIEW

**Comment by steerben** [comment]
Would be really cool if this could be merged. It's production critical for us

**Comment by yovanoc** [comment]
I have a problem with both request and axios.

image
image

But how a status code 200 can contain a bad request??? thanks

**Comment by stoneChen** [comment]
@yovanoc 
Is this problem happened due to proxy?
If not, the log line "data HTTP/1.1 ..." is the body of the request, not the response header, maybe something is wrong with the server.
Additionally, I try blew:
```js
const request = require('request')
request({
  json: true,
  timeout: 10000,
  tunnel: false,
  url: 'https://haapi.ankama.com/json/Ankama/v2/Account/CreateGuest?game=20&lang=fr',
}, function (error, response, body) {
  console.log('error:', error);
  console.log('statusCode:', response && response.statusCode);
  console.log(body)
});
```
I got:
```
error: null
statusCode: 503
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
....
```

**Comment by yovanoc** [comment]
oh weird because for me it doesn't work only with proxy .. and I've written before a c# program that does the same and all was working without pain

**Comment by Sitronik** [comment]
Unfortunately, this pull request doesn't merged, but I fixed this issue too https://github.com/Sitronik/axios

**Comment by ddgenome** [comment]
Until this gets merged and released, here a published version of this fix: https://atomist.jfrog.io/atomist/npm-dev/axios/-/axios-0.17.1-proxy-fix-20180111160124.tgz

You can use this version of axios in you project using the following command:

```
$ npm install --save https://atomist.jfrog.io/atomist/npm-dev/axios/-/axios-0.17.1-proxy-fix-20180111160124.tgz
```

**Comment by ddgenome** [comment]
We've have better luck with the fix by @Sitronik https://github.com/axios/axios/compare/master...Sitronik:master

Here's a published version of that https://atomist.jfrog.io/atomist/npm-dev/axios/-/axios-0.17.1-proxy-fix.tgz

File: package.json [hunk 1]
```diff
@@ -71,8 +71,9 @@
   },
   "typings": "./index.d.ts",
   "dependencies": {
+    "is-buffer": "^1.1.5",
```
Comment (yasuf) [inline package.json]: please remove dependency, we avoided it since it's so small we can just have the function here

File: lib/adapters/http.js [hunk 2]
```diff
@@ -116,16 +116,40 @@ module.exports = function httpAdapter(config) {
...
+          ca: options.ca,
+          cert: options.cert,
+          key: options.key,
+          passphrase: options.passphrase,
+          pfx: options.pfx,
+          ciphers: options.ciphers,
+          rejectUnauthorized: options.rejectUnauthorized,
+          secureOptions: options.secureOptions,
+          secureProtocol: options.secureProtocol
+        });
```
Comment (yasuf) [inline lib/adapters/http.js]: this adapter is pretty big, sort of tricky to follow the logic, would you mind breaking the logic down into methods? my main concern is having deeply nested `if` conditionals, it's sort of hard to ensure our logic is sound with these

File: test/unit/adapters/http.js [hunk 3]
```diff
@@ -330,6 +330,37 @@ module.exports = {
...
+          port: 4000
+        }
+      }).then(function (res) {
+        success = true
+        test.equal(res.status, 200, 'should pass through proxy');
+        test.done();
+      })
+    });
+  },
+
```
Comment (yasuf) [inline test/unit/adapters/http.js]: thanks for adding a test!

**Comment by jasonsaayman** [comment]
Hi,

Thank you for this pull request, however it seem that it has gone stale and is either far out of date with the current master head or has unresolved review issues. Should you feel that this pull request is still relevant please submit a new pull request.

Thanks

**Pull request closed without merging**